File: micropython/examples/breakout_dotmatrix/bargraph.py
import time
import math
import random

from pimoroni_i2c import PimoroniI2C
from breakout_dotmatrix import BreakoutDotMatrix

PINS_BREAKOUT_GARDEN = {"sda": 4, "scl": 5}
PINS_PICO_EXPLORER = {"sda": 20, "scl": 21}

i2c = PimoroniI2C(**PINS_BREAKOUT_GARDEN)
display = BreakoutDotMatrix(i2c)
width = display.WIDTH
height = display.HEIGHT

values = [0, 0, 0, 0, 0]
start_time = time.time()


while True:
    # Add a new random value to our list and prune the list to visible values
    values.insert(0, random.randint(0, height))
    values = values[:width]

    # Animate a value from 0 to height + 1
    value = (math.sin((time.time() - start_time) * math.pi) + 1) / 2.0
    value *= height + 1
    value = math.floor(value)

    for y in range(height):
        y = height - 1 - y

        for x in range(width // 2):
            # Left
            display.set_pixel(x, y, value <= y)

            # Right
            display.set_pixel(x + (width // 2), y, values[x] <= y)

    display.show()
    time.sleep(1.0 / height)

File: examples/breakout_dotmatrix/bargraph/bargraph.cpp
#include <math.h>

#include "breakout_dotmatrix.hpp"
#include "time.h"

using namespace pimoroni;

I2C i2c(BOARD::BREAKOUT_GARDEN);
BreakoutDotMatrix display(&i2c);
static const uint8_t WIDTH = display.WIDTH;
static const uint8_t HEIGHT = display.HEIGHT;
static const uint8_t HALF_WIDTH = (WIDTH / 2);

uint8_t values[HALF_WIDTH] = { 0, 0, 0, 0, 0 };
uint8_t next_value = 0;

int main() {
  gpio_init(PICO_DEFAULT_LED_PIN);
  gpio_set_dir(PICO_DEFAULT_LED_PIN, GPIO_OUT);

  display.init();

  while(true) {
    gpio_put(PICO_DEFAULT_LED_PIN, true);

    absolute_time_t at = get_absolute_time();
    uint64_t t = to_us_since_boot(at) / 1000000;

    // Add a new random value to our array, overriding the oldest value
    values[next_value] = rand() % HEIGHT;
    next_value = (next_value + 1) % HALF_WIDTH;

    // Animate a value from 0 to height + 1
    float value = (sinf(t * M_PI) + 1) / 2.0f;
    value *= HEIGHT + 1;
    value = floor(value);

    for(uint8_t y = 0; y < HEIGHT; y++) {
      uint8_t y_inv = HEIGHT - 1 - y;

      for(uint8_t x = 0; x < HALF_WIDTH; x++) {
        // Left
        display.set_pixel(x, y_inv, value <= y_inv);

        // Right
        display.set_pixel(x + HALF_WIDTH, y_inv, values[(x + next_value) % HALF_WIDTH] <= y_inv);
      }
    }

    display.show();

    gpio_put(PICO_DEFAULT_LED_PIN, false);

    sleep_ms(1000 / HEIGHT);
  }

  return 0;
}